Snakemake workflow

rnmitchell opened the pull request:
This PR will create a snakemake workflow to simplify running the entire lusSTR pipeline on either a single file or a set of files.

standage commented on lusSTR/workflow/.DS_Store:
Nope.

standage commented:
No concerns about the Snakefile at the moment. I'd be happy to chat about some streamlining ideas some time, but that would be icing on the cake.

rnmitchell commented:
Sounds good!

rnmitchell commented on lusSTR/cli/config.py:
```diff
@@ -0,0 +1,106 @@
+import yaml
+
+
+def main(args):
+    Path(args.workdir).mkdir(parents=True, exist_ok=True)
+    final_dest = f"{args.workdir}/config.yaml"
+    config = resource_filename("lusSTR", "data/config.yaml")
+    final_config = edit_config(config, args)
+    with open(final_dest, "w") as file:
+        yaml.dump(final_config, file)
+
```
I created the command ```lusstr config``` to create the config file. Settings can be specified as command line arguments. If no arguments are provided, the default settings are used and the user can edit manually. If a working directory is specified, the config file is dumped there, otherwise to the cwd.

rnmitchell commented:
This is ready for further review @standage. I added a new command (```lusstr config```) to create the config file to store all the settings for the STR workflow, allowing the user to specify settings via the command line or they could open the file once created and manually edit it (probably the preferred way for non-bioinformatics folks 😄 ). The entire workflow doesn't have to be run (can just run up to the ```annotate``` step or just the ```format``` step if wanted).  

I also updated all tests and all are passing with the exception of ```test_snps.py```, as I haven't updated any of the snp workflow at this time. Whenever you have some time to poke around and give any advice and try out the tests, that would be great!

rnmitchell commented:
> I do still need to update the README.

Nevermind! updated!

standage commented on lusSTR/tests/test_suite.py:
```diff
@@ -136,83 +147,143 @@ def test_indel_flag(locus, sequence, uas, kit, output):
+        ("all", "lusstr_output/Positive_Control_evidence_ngs.csv", True, True, True),
+    ],
+)
+def test_snakemake(command, output, format_out, annot_out, all_out, tmp_path):
+    config = str(tmp_path / "config.yaml")
+    inputfile = data_file("UAS_bulk_input/Positive Control Sample Details Report 2315.xlsx")
+    exp_output = data_file(output)
```
Running this data through `lusstr strs all` on the command line went seamlessly for me.

standage reviewed (commented):
Overall, this looks pretty good. A few comments, but hopefully should be quick to resolve.

One thing you might consider is that the various lusstr subcommands give users the flexibility to run the STR analysis workflow step-wise, so having to specify format, annotate, or all seems a bit redundant. Why is the user going to run `lusstr strs` if they don't want to run `all`? There might be a good reason for this, and I could probably formulate the contours of that reason myself. But I'm curious what you think. If we cut out that argument, that could simplify the interface a bit (I'm not sure it would simplify the workflow implementation at all). If you decide that it's necessary and should stay, how would you feel about making it an option with `all` as the default?

rnmitchell commented on lusSTR/cli/config.py:
```diff
@@ -0,0 +1,106 @@
+        data["separate"] = True
+    if args.nocombine:
+        data["nocombine"] = True
+    if args.efm_sep:
```
Nope, old argument that I forgot to remove from that function. There were two different ```separate``` pieces, one for the ```annotate``` step and one for the ```filter``` step. Now only the ```filter``` step argument remains.

rnmitchell commented:
So do you mean by running ```lusstr format``` or ```lusstr annotate```? Those subcommands no longer work since I refactored all the code for snakemake.

rnmitchell commented:
I believe this is ready now, @standage!

standage reviewed (approved):
LGTM!

standage commented:
It looks like some `filter` tests are failing due to the DataFrame `append` method not being found. I noted that pandas 2.0 was released recently and probably has some breaking changes. We need to pin both minimum and maximum version numbers for pandas now, it looks like: `pandas>=1.0,<2.0`.

SNP tests failed as well. If you're saving the SNP workflow for another PR, you should probably disable those tests for now.

rnmitchell commented:
Ok will look into this!

rnmitchell commented:
Ok looks like things are passing now! Everything good for you?
